src/progress/progress_set.rs

```diff
@@ -369,21 +368,49 @@ impl ProgressSet {
         );
     }
 
-    /// Returns the maximal committed index for the cluster.
+    /// Returns the maximal committed index for the cluster. The bool flag indicates whether
+    /// the index is computed by group commit algorithm successfully.
     ///
     /// Eg. If the matched indexes are [2,2,2,4,5], it will return 2.
-    pub fn maximal_committed_index(&self) -> u64 {
-        let mut matched = self.sort_buffer.borrow_mut();
+    /// If the matched indexes and groups are `[(1, 1), (2, 2), (3, 2)]`, it will return 1.
+    pub fn maximal_committed_index(&mut self) -> (u64, bool) {
+        let matched = &mut self.sort_buffer;
         matched.clear();
+        let mut use_group_commit = false;
+        let progress = &self.progress;
         self.configuration.voters().iter().for_each(|id| {
-            let peer = &self.progress[id];
-            matched.push(peer.matched);
+            let p = &progress[id];
+            use_group_commit |= p.commit_group_id != 0;
+            matched.push((p.matched, p.commit_group_id));
         });
         // Reverse sort.
-        matched.sort_by(|a, b| b.cmp(a));
+        matched.sort_by(|a, b| b.0.cmp(&a.0));
 
         let quorum = crate::majority(matched.len());
-        matched[quorum - 1]
+        if !use_group_commit {
+            return (matched[quorum - 1].0, false);
+        }
+        let (quorum_commit_index, mut checked_group_id) = matched[quorum - 1];
+        let mut single_group = true;
+        for (index, group_id) in matched.iter() {
+            if *group_id == 0 {
+                single_group = false;
+                continue;
+            }
+            if checked_group_id == 0 {
+                checked_group_id = *group_id;
+                continue;
+            }
+            if checked_group_id == *group_id {
+                continue;
+            }
+            return (cmp::min(*index, quorum_commit_index), true);
+        }
+        if single_group {
+            (matched[quorum - 1].0, false)
+        } else {
+            (matched.last().unwrap().0, false)
+        }
```

Comment:
I think this algorithm can't calculate the max commit index.
If the progress is (1, 1) (2, 2) (3, 1), this algorithm's answer is 1 but the correct answer is 2.